sdk/core/azure-core/src/base64.cpp

```diff
@@ -313,10 +313,10 @@ static void Base64WriteIntAsFourBytes(char* destination, int32_t value)
   destination[0] = static_cast<uint8_t>(value & 0xFF);
 }
 
-std::string Base64Encode(const std::vector<uint8_t>& data)
+std::string Base64Encode(uint8_t const* const data, size_t length)
 {
   size_t sourceIndex = 0;
-  auto inputSize = data.size();
+  auto inputSize = length;
```

Comment:
These changes are necessary if we want to be able to base64 encode data without taking trips through the heap allocator.

The global namespace Base64Encode function should operate on a pointer and length, while callers of Base64Encode should take whatever their encoding functions is and pass the raw data to the global Base64Encode function.

So we would have:

```cpp
std::string Base64Encode(std::string const& input)
{
   return ::Base64Encode(input.data(), input.size());
}
```

and
```cpp
std::string Base64Encode(std::vector<uint8_t> const& input)
{
  return ::Base64Encode(input.data(), input.size());
}
```

This way we can base64 encode strings and vectors *without* making a round trip through the heap to allocate a temporary `std::vector<uint8_t>`.

Reply:
Adding: At one point in this pull request, I had exactly those functions. I no longer need them, however the underlying refactorings to avoid additional allocations are still righteous changes.